graphql_client_codegen/src/lib.rs

```diff
@@ -43,83 +44,122 @@ impl Display for GeneralError {
-        .and_then(std::ffi::OsStr::to_str)
-        .unwrap_or("INVALID");
-    let schema_string;
-
-    // Check the schema cache.
-    let schema: schema::Schema = {
-        let mut lock = SCHEMA_CACHE.lock().expect("schema cache is poisoned");
-        match lock.entry(schema_path.to_path_buf()) {
+    let cached: V = {
+        let mut lock = cache.lock().expect("cache is poisoned");
+        match lock.entry(key) {
             btree_map::Entry::Occupied(o) => o.get().clone(),
-            btree_map::Entry::Vacant(v) => {
-                schema_string = read_file(v.key())?;
-                let schema = match schema_extension {
-                    "graphql" | "gql" => {
-                        let s = graphql_parser::schema::parse_schema::<&str>(&schema_string).map_err(|parser_error| GeneralError(format!("Parser error: {}", parser_error)))?;
-                        schema::Schema::from(s)
-                    }
-                    "json" => {
-                        let parsed: graphql_introspection_query::introspection_response::IntrospectionResponse = serde_json::from_str(&schema_string)?;
-                        schema::Schema::from(parsed)
-                    }
-                    extension => return Err(GeneralError(format!("Unsupported extension for the GraphQL schema: {} (only .json and .graphql are supported)", extension)).into())
-                };
-
-                v.insert(schema).clone()
-            }
+            btree_map::Entry::Vacant(v) => v.insert(value).clone(),
         }
     };
 
-    // We need to qualify the query with the path to the crate it is part of
-    let (query_string, query) = {
-        let mut lock = QUERY_CACHE.lock().expect("query cache is poisoned");
-        match lock.entry(query_path) {
-            btree_map::Entry::Occupied(o) => o.get().clone(),
-            btree_map::Entry::Vacant(v) => {
-                let query_string = read_file(v.key())?;
-                let query = graphql_parser::parse_query(&query_string)
-                    .map_err(|err| GeneralError(format!("Query parser error: {}", err)))?
-                    .into_static();
-                v.insert((query_string, query)).clone()
+    Ok(cached)
+}
+
+fn query_document(query_string: &str) -> Result<QueryDocument, BoxError> {
+    let document = graphql_parser::parse_query(query_string)
+        .map_err(|err| GeneralError(format!("Query parser error: {}", err)))?
+        .into_static();
+    Ok(document)
+}
+
+fn get_set_query_from_file(
+    query_path: &std::path::Path,
+) -> Result<(String, QueryDocument), BoxError> {
+    get_set_cached(&QUERY_CACHE, query_path.to_path_buf(), {
+        let query_string = read_file(query_path)?;
+        let query_document = query_document(&query_string)?;
+        (query_string, query_document)
+    })
+}
+
+fn get_set_schema_from_file(schema_path: &std::path::Path) -> Result<Schema, BoxError> {
+    get_set_cached(&SCHEMA_CACHE, schema_path.to_path_buf(), {
```

Comment:
Tiny DX hack :D This allows these kind of functions to be called with a `PathBuf`, `&PathBuf`, a `&Path` or even a `&&Path`. It can simplify some of the code further below. Feel free to dismiss this altogether tho.

```suggestion
fn get_set_schema_from_file(schema_path: impl Into<PathBuf>) -> Result<Schema, BoxError> {
    let schema_path = schema_path.into();
     get_set_cached(&SCHEMA_CACHE, schema_path.clone(), {
```

Reply:
`PathBuf` implements `Deref<Target = Path>` so except for the owned type, everything should already work.

Reply:
Ah yeah, good point. I was mostly going by the fact that I saw a `.as_path()` further down.

Reply:
👍🏽 I passed `schema_path` here and used `key.into()`